### PR TITLE
feat: add dynamic_programming/subset_sum.cpp  (subset sum problem)

### PR DESCRIPTION
**Pull request by Vinay-Garg**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1434"></a>

### CODE REVIEW

File: dynamic_programming/subset_sum.cpp [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+ * Output: false
+ * Explanation: There is no subset that add up to 30.
+ *
+ * @author [Vinay Garg] (https://github.com/Vinay-Garg)
+ */
+
+#include <array>
+#include <cassert>
+#include <iostream>
+#include <vector>
```
Comment (Panquesito7) [inline dynamic_programming/subset_sum.cpp:31]:
```suggestion
#include <array>       /// for std::array
#include <cassert>    /// for assert
#include <iostream>   /// for io operations
#include <vector>     /// for std::vector
```

File: dynamic_programming/subset_sum.cpp [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic programming algorithms
+ */
+namespace dynamic_programming {
+
+/**
+ * @namespace subset_sum_bottomup
+ * @brief bottom up implementation of subset sum problem
```
Comment (Panquesito7) [inline dynamic_programming/subset_sum.cpp:41]:
```suggestion
* @brief Bottom up implementation of [Subset Sum Problem](https://en.wikipedia.org/wiki/Subset_sum_problem)
```

File: dynamic_programming/subset_sum.cpp [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+/**
+ * @brief This is bottom up implemenation for checking if a subset 
+ * exists for the given sum value
+ * @tparam n 	: size of the set array
+ * @param set 	: an array of numbers in set of size<=n
+ * @param sum 	: subset sum value
+ * @returns true if there is a subset of set with sum equal to given sum
+ */
+template <size_t n>
+bool subset_sum(const std::array<int, n> &set, const int sum) {
```
Comment (Panquesito7) [inline dynamic_programming/subset_sum.cpp:53]:
```suggestion
bool subset_sum(const std::array<int, n> &set, const int& sum) {
```

File: dynamic_programming/subset_sum.cpp [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+	bool ans = subset[n][sum];
+	return ans;
+} 
+
+}  // namespace subset_sum_bottomup
+
+
+/**
+ * @namespace subset_sum_topdown
+ * @brief top down implementation of subset sum problem
```
Comment (Panquesito7) [inline dynamic_programming/subset_sum.cpp:89]:
```suggestion
 * @brief Top down implementation of [Subset Sum Problem](https://en.wikipedia.org/wiki/Subset_sum_problem)
```

File: dynamic_programming/subset_sum.cpp (hunk 3 above)
Comment (Panquesito7) [inline dynamic_programming/subset_sum.cpp:53]: Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)).

File: dynamic_programming/subset_sum.cpp [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+ */
+namespace subset_sum_topdown {
+/**
+ * @brief This is top down implemenation is for checking if a subset 
+ * exists for the given sum value
+ * @param T template-type to use any kind of value
```
Comment (Panquesito7) [inline dynamic_programming/subset_sum.cpp:95]:
```suggestion
 * @tparam T template-type to use any kind of value
```

File: dynamic_programming/subset_sum.cpp [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+ * @namespace subset_sum_bottomup
+ * @brief bottom up implementation of subset sum problem
+ */
+namespace subset_sum_bottomup {
+/**
+ * @brief This is bottom up implemenation for checking if a subset 
+ * exists for the given sum value
+ * @tparam n 	: size of the set array
+ * @param set 	: an array of numbers in set of size<=n
+ * @param sum 	: subset sum value
```
Comment (Panquesito7) [inline dynamic_programming/subset_sum.cpp:49]:
```suggestion
 * @tparam n size of the set array
 * @param set an array of numbers in set of size<=n
 * @param sum subset sum value
```

File: dynamic_programming/subset_sum.cpp [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+
+    // exclude last element of set in vector v and
+	// search for sunset with sum value equals sum
+    bool exc =  subset_sum(set, n - 1, sum,dp);
+
+    return dp[n][sum] = inc || exc;
+}
+
+}  // namespace subset_sum_topdown
+
```
Comment (Panquesito7) [inline dynamic_programming/subset_sum.cpp:130]:
```suggestion
```

File: dynamic_programming/subset_sum.cpp [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+ * @brief top down implementation of subset sum problem
+ */
+namespace subset_sum_topdown {
+/**
+ * @brief This is top down implemenation is for checking if a subset 
+ * exists for the given sum value
+ * @param T template-type to use any kind of value
+ * @param set 	: an array of numbers in set of size<=n
+ * @param n 	: size of the set array
+ * @param sum 	: subset sum value
```
Comment (Panquesito7) [inline dynamic_programming/subset_sum.cpp:98]:
```suggestion
 * @param set an array of numbers in set of size<=n
 * @param n size of the set array
 * @param sum subset sum value
```

File: dynamic_programming/subset_sum.cpp [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+              << std::endl;
+
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    test();  // call the test function :)
```
Comment (Panquesito7) [inline dynamic_programming/subset_sum.cpp:193]: Adding `return 0;` at the end is a good programming practice. 🙂
```suggestion
    test();  // run self-test implementations
    return 0;
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
👍 Well done!

**Review by Vinay-Garg — COMMENTED** [review]
Thanks for the suggestions.

File: dynamic_programming/subset_sum.cpp [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+/**
+ * @brief This is bottom up implemenation for checking if a subset 
+ * exists for the given sum value
+ * @tparam n size of the set array
+ * @param set an array of numbers in set of size<=n
+ * @param sum subset sum value
+ * @returns true if there is a subset of set with sum equal to given sum
+ */
+template <size_t n>
+bool subset_sum(const std::array<int, n> &set, const int& sum) {
```
Comment (Panquesito7) [inline dynamic_programming/subset_sum.cpp:53]: Consider using `uint64_t`, `uint32_t`, `uint16_t` or `uint8_t` for non-negative values. For negative values, use `int64_t` or their appropriate size. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)).

**Review by Panquesito7 — COMMENTED** [review]
Good work! 👍

**Pull request closed without merging**